### PR TITLE
Add tests for the Swift snippets workflows

### PR DESCRIPTION
**Pull request by award999**
* Created some common utilities for dealing with debug sessions
* Add integration test for run command and debug variant, big difference
  being whether the breakpoint is hit or not

Issue: #1125

### CODE REVIEW

File: test/integration-tests/SwiftSnippet.test.ts [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+        // Set a breakpoint
+        vscode.debug.addBreakpoints(breakpoints);
+    });
+
+    suiteTeardown(async () => {
+        await vscode.commands.executeCommand("workbench.action.closeAllEditors");
+        vscode.debug.removeBreakpoints(breakpoints);
+    });
+
+    test("Run `Swift: Run Swift Snippet` command for snippet file", async () => {
```
Comment (award999) [inline test/integration-tests/SwiftSnippet.test.ts:50]: Mark as `@slow before merging, just want to make sure passes in CI

File: test/integration-tests/SwiftSnippet.test.ts [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+
+        const session = await sessionPromise;
+        expect(session.configuration).to.have.property(
+            "program",
+            `${testAssetPath("defaultPackage")}/.build/debug/hello`
+        );
+        expect(session.configuration).to.have.property("noDebug", true);
+    }).timeout(120000);
+
+    test("Run `Swift: Debug Swift Snippet` command for snippet file", async () => {
```
Comment (award999) [inline test/integration-tests/SwiftSnippet.test.ts:67]: Mark as `@slow before merging, just want to make sure passes in CI

**Comment by michael-weng** [comment]
The CI will probably make this apparent so might just be stating the obvious here, but snippet's only implemented in 5.7 so probably need to set some precondition.
https://github.com/swiftlang/swift-evolution/blob/main/proposals/0356-swift-snippets.md

**Comment by award999** [comment]
> The CI will probably make this apparent so might just be stating the obvious here, but snippet's only implemented in 5.7 so probably need to set some precondition. https://github.com/swiftlang/swift-evolution/blob/main/proposals/0356-swift-snippets.md

yep thanks :), also seems like it was hidden behind an experimental flag for a bit so trying to run that down

**Comment by michael-weng** [comment]
> > The CI will probably make this apparent so might just be stating the obvious here, but snippet's only implemented in 5.7 so probably need to set some precondition. https://github.com/swiftlang/swift-evolution/blob/main/proposals/0356-swift-snippets.md
> 
> yep thanks :), also seems like it was hidden behind an experimental flag for a bit so trying to run that down

Nice! If it's a particular flag we are expecting would be a good candidate for a separate contract test, and if anything were to change we will know explicit the contract with the driver's broken.

File: src/SwiftSnippets.ts [hunk 3]
```diff
@@ -87,24 +87,21 @@ export async function debugSnippetWithOptions(
...
-                        folderContext.workspaceFolder,
-                        snippetDebugConfig,
-                        options
-                    );
-                }
-            });
-    } catch {
-        // ignore error if task failed to run
-    }
+    // try {
```
Comment (michael-weng) [inline src/SwiftSnippets.ts:90]: Forgot to clean up?

File: test/integration-tests/SwiftSnippet.test.ts [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+
+suite("SwiftSnippet Test Suite", () => {
+    const uri = testAssetUri("defaultPackage/Snippets/hello.swift");
+    const breakpoints = [
+        new vscode.SourceBreakpoint(new vscode.Location(uri, new vscode.Position(2, 0))),
+    ];
+
+    suiteSetup(async function () {
+        const folder = await folderContextPromise("defaultPackage");
+        if (folder.workspaceContext.toolchain.swiftVersion.isLessThan(new Version(6, 0, 0))) {
```
Comment (michael-weng) [inline test/integration-tests/SwiftSnippet.test.ts:36]: IIRC the source specified 5.7 for snippet feature, something in the CI hindering us I guess?

File: src/SwiftSnippets.ts (hunk 3 above)
Comment (award999) [reply src/SwiftSnippets.ts:90]: yep 👍

File: test/integration-tests/SwiftSnippet.test.ts [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+        const exitCode = await exitPromise;
+        expect(exitCode).to.equal(0);
+
+        const session = await sessionPromise;
+        expect(session.configuration).to.have.property(
+            "program",
+            `${testAssetPath("defaultPackage")}/.build/debug/hello`
+        );
+        expect(session.configuration).to.have.property("noDebug", true);
+    }).timeout(120000);
```
[the remark this answers is not in the document]
Comment (award999) [reply test/integration-tests/SwiftSnippet.test.ts:75]: done

File: test/utilities/debug.ts [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+                res(e);
+            }
+        })
+    );
+}
+
+export async function waitForDebugAdapterMessage(
+    name: string,
+    matches: (message: any) => boolean
+): Promise<any> {
```
[the remark this answers is not in the document]
Comment (award999) [reply test/utilities/debug.ts:32]: Screenshot 2024-12-10 at 12 35 22 PM
 that is what is in the VSCode API types. I could add https://www.npmjs.com/package/@vscode/debugprotocol as a dependency

[the remark this answers is not in the document]
Comment (award999) [reply test/utilities/debug.ts]: done

File: test/utilities/debug.ts [hunk 7]
```diff
@@ -65,14 +75,28 @@ export async function waitForDebugAdapterMessage(
...
+}
+
+/**
+ * Waits for a debug adapter for the specified DebugSession name to terminate.
+ *
+ * @param name The name of the debug session to wait for.
+ * @returns exit code of the DAP
+ */
+export async function waitForDebugAdapterExit(name: string): Promise<number> {
+    return await waitForDebugAdapterEvent(name, "exited").then(m => m.body.exitCode);
```
[the remark this answers is not in the document]
Comment (award999) [reply test/utilities/debug.ts]: done

File: test/integration-tests/SwiftSnippet.test.ts (hunk 1 above)
Comment (award999) [reply test/integration-tests/SwiftSnippet.test.ts:72]: Done

File: test/integration-tests/SwiftSnippet.test.ts (hunk 2 above)
Comment (award999) [reply test/integration-tests/SwiftSnippet.test.ts:85]: done

**Review by plemarquand — APPROVED** [review]

**Pull request merged**